# 아이템 39 : 명명 패턴보다 애너테이션을 사용하라


전통적으로 도구나 프레임워크가 특별히 다뤄야 할 프로그램 요소에는 딱 구분되는 **명명 패턴**을 적용해왔다.

- 테스트 프레임워크인 JUnit은 버전 3까지 테스트 메서드 이름을 test로 시작하게끔 했다.

```java
public class HelloTest extends TestCase {
	public void testGetUsers() {
		return getUsers();
	}
}
```

그러나 이러한 명명패턴은 몇가지 단점을 가지고 있다.

### 첫 번째, 오타가 나면 안 된다.

### 두 번째, 올바른 프로그램 요소에서만 사용 되리라 보증할 방법이 없다.

- 예컨대 (메서드가 아닌)클래스 이름을 `TestSafetyMechanism`로 지어 `JUnit`에 던져 줬다고 하면, 개발자는 이 클래스에 정의된 테스트 메서드들을 수행해주길 기대하겠지만, `JUnit`은 클래스 이름에는 관심이 없다.
- `JUnit`은 경고 메시지조차 출력하지 않으면서 의도한 테스트가 전혀 수행되지 않을 것이다.

### 세 번째, 프로그램 요소를 매개 변수로 전달할 마땅한 방법이 없다.

- 특정 예외를 던져야만 성공하는 테스트가 있다고 가정 했을 때,
  기대하는 예외 타입을 테스트에 매개변수로 전달해야 하는 상황이다.
- 예외의 이름을 테스트 메서드 이름에 덧붙이는 방법도 있지만 보기도 나쁘고 깨지기도 쉽다.
- 컴파일러는 메서드 이름에 덧붙인 문자열이 예외를 가르키는 지 알 도리가 없다.

에너테이션은 이 모든 문제를 해결해주는 개념으로 `Junit`도 버전 4부터 전면 도입하였다.

> 새롭게 제작한 작은 테스트 프레임워크
>

```java
/**
 * 테스트 메서드임을 선언하는 애너테이션이다.
 * 매개변수 없는 정적 메서드 전용이다.
 */
@Retention(RetentionPolicy.RUNTIME) // @Test가 런타임에도 유지되어야 한다는 표시이다. 
@Target(ElementType.METHOD) // @Test가 반드시 메서드 선언에서만 사용돼야 한다. 
public @interface Test { // @Test
}
```

- “매개변수 없는 정적 메서드 전용”이라고 쓰여있지만, 이 제약을 컴파일러가 강제하기 위해서는 적절한 애너테이션 처리기를 직접 구현해야 한다.
- 관련 방법은 `javax.annotation.processing` API문서를 참고
    - 적절한 애너테이션 처리기 없이 인스턴스 메서드나 매개변수가 있는 메서드에 달면?
        - 컴파일은 잘되지만 테스트 도구를 실행할 때 문제가 된다.

> 이러한 애너테이션을 아무 매개변수 없이 단순히 대상에 **마킹**한다는 뜻에서 마커 애너테이션이라 한다.
>

**마커 애너테이션을 동작하는 프로그램 예시 코드**

```java
public class RunTests {
    public static void main(String[] args) throws Exception {
        int tests = 0;
        int passed = 0;
        Class<?> testClass = Class.forName(args[0]); // 완전 정규화된 클래스 이름
        for (Method m : testClass.getDeclaredMethods()) {
            if (m.isAnnotationPresent(Test.class)) { // 실행할 메서드를 찾는다.
                tests++;
                try {
                    m.invoke(null);
                    passed++;
                } catch (InvocationTargetException wrappedExc) { // 테스트 메서드가 예외를 던지면 여기서 잡는다.
                    Throwable exc = wrappedExc.getCause();
                    System.out.println(m + " 실패: " + exc);
                } catch (Exception exc) { // @Test 애너테이션을 잘못 사용한 경우 여기서 잡는다.
                                          // 인스턴스 메서드, 매개변수가 있는 메서드, 호출할 수 없는 메서드 등에 달은 경우
                    System.out.println("잘못 사용한 @Test: " + m);
                }
            }
        }
        System.out.printf("성공: %d, 실패: %d%n",
                passed, tests - passed);
    }
}
```

이 테스트 러너는 명령줄로부터 완전 정규화된 클래스 이름을 받아,

그 클래스에서 `@Test` 에너테이션이 달린 메서드를 차례로 호출한다.

다음은 특정 예외를 던져야만 성공하는 테스트를 지원하도록 해보자.

```jsx
/**
 * 명시한 예외를 던져야만 성공하는 테스트 메서드용 애너테이션
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface ExceptionTest {
    Class<? extends Throwable> value();
}
```

- `Class<? extends Throwable>` : Throwable을 확장한 클래스의 Class 객체
    - 모든 예와 오류 타입을 다 수용한다.
- `@ExceptionTest(ArithmeticException.class)`와 같이 사용

**`@ExceptionTest`** 애너테이션을 다룰 수 있도록 아래와 같이 테스트 도구를 수정

```jsx
public class RunTests {
    public static void main(String[] args) throws Exception {
        int tests = 0;
        int passed = 0;
        Class<?> testClass = Class.forName(args[0]);
        for (Method m : testClass.getDeclaredMethods()) {
            if (m.isAnnotationPresent(Test.class)) {
                tests++;
                try {
                    m.invoke(null);
                    passed++;
                } catch (InvocationTargetException wrappedExc) {
                    Throwable exc = wrappedExc.getCause();
                    System.out.println(m + " 실패: " + exc);
                } catch (Exception exc) {
                    System.out.println("잘못 사용한 @Test: " + m);
                }
            }
            
						// 추가된 부분
						// @Test 애너테이션용 코드와 비슷하지만, 
						// 이 코드는 애너테이션 매개변수의 값을 추출하여 테스트 메서드가 올바른 예외를 던지는지 확인하는데 사용한다
            if (m.isAnnotationPresent(ExceptionTest.class)) {
                tests++;
                try {
                    m.invoke(null);
                    System.out.printf("테스트 %s 실패: 예외를 던지지 않음%n", m);
                } catch (InvocationTargetException wrappedEx) {
                    Throwable exc = wrappedEx.getCause();
                    Class<? extends Throwable> excType =
                            m.getAnnotation(ExceptionTest.class).value();
                    if (excType.isInstance(exc)) {
                        passed++;
                    } else {
                        System.out.printf(
                                "테스트 %s 실패: 기대한 예외 %s, 발생한 예외 %s%n",
                                m, excType.getName(), exc);
                    }
                } catch (Exception exc) {
                    System.out.println("잘못 사용한 @ExceptionTest: " + m);
                }
            }
        }

        System.out.printf("성공: %d, 실패: %d%n",
                passed, tests - passed);
    }
}
```

```jsx
public class Sample2 {
    @ExceptionTest(ArithmeticException.class)
    public static void m1() {  // 성공해야 한다.
        int i = 0;
        i = i / i;
    }
    @ExceptionTest(ArithmeticException.class)
    public static void m2() {  // 실패해야 한다. (다른 예외 발생)
        int[] a = new int[0];
        int i = a[1];
    }
    @ExceptionTest(ArithmeticException.class)
    public static void m3() { }  // 실패해야 한다. (예외가 발생하지 않음)
}
```

이 예외 테스트 예에서 한 걸음 더 들어가, 예외를 여러 개 명시하고 그중 하나가 발생하면 성공하게 만들 수도 있다.

```jsx
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface ExceptionTest {
    Class<? extends Exception>[] value();
}
```

배열 매개변수를 받는 애너테이션용 문법은 아주 유연하다. 단일 원소 배열에 최적화했지만, 앞서의 `@ExceptionTest`들도 모두 수정 없이 수용한다.

```jsx
@ExceptionTest({IndexOutOfBoundsException.class, NullPointerException.class})
public static void doublyBad() {   // 성공해야 한다.
    List<String> list = new ArrayList<>();

    // 자바 API 명세에 따르면 다음 메서드는 IndexOutOfBoundsException이나
    // NullPointerException을 던질 수 있다.
    list.addAll(5, null);
}
```

1. **IndexOutOfBoundsException**: **`list`** 가 빈 리스트이므로 유효한 인덱스는 0 뿐입니다. 그러나 **`5`** 번째 인덱스에 요소를 추가하려고 하면 **`IndexOutOfBoundsException`** 이 발생
2. **NullPointerException**: **`null`** 값을 전달하면 **`addAll`** 메서드는 내부적으로 **`null`** 이 아닌 컬렉션을 기대하므로 **`NullPointerException`** 이 발생

> **자바 API 명세에서는 index 먼저 검사하기 때문에 IndexOutOfBoundsException이 뜨는데, 혹시 5라는 Index가 존재할수도있으니까 NullPointerException은 보혐용으로 추가해 놓은 것이라 판단된다.**
>

자바 8에서는 여러 개의 값을 받는 애너테이션을 다른 방식으로도 만들 수 있다.
배열 매개변수를 사용하는 대신 애너테이션에 `@Repeatable` 메타 애너테이션을 다는 방식이다. `@Repeatable`을 단 애너테이션은 하나의 프로그램 요소에 여러 번 달 수 있다.

**@Repeatable은 주의해야 할점이 있다.**

- (1) `Repeatable`을 단 애너테이션을 반환하는 '컨테이너 애너테이션'을 하나 더 정의하고, `@Repeatable`에 이 컨테이너 애너테이션의 class 객체를 매개변수로 전달해야 한다.
- (2) 컨테이너 애너테이션은 내부 애너테이션 타입의 배열을 반환하는 value 메서드를 정의해야 한다.
- (3) 컨테이너 에너테이션 타입에는 적절한 보존 정책(`@Retention`)과 적용 대상(`@Target`)을 명시해야 한다. → 그렇지 않으면 컴파일 되지 않는다.

```jsx
// 반복 가능한 애너테이션
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Repeatable(ExceptionTestContainer.class) // -> (1)
public @interface ExceptionTest {
    Class<? extends Throwable> value();
}

// 컨테이너 애너테이션
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)                // -> (3)
public @interface ExceptionTestContainer {
		ExceptionTest[] value();              // -> (2)
}
```

**반복가능 애너테이션은 처리할 때에도 주의를 요한다.**

반복 가능 애너테이션을 여러 개 달면 하나만 달았을 때와 구분하기 위해 해당 **컨테이너 애너테이션 타입이 적용된다.**

**(1) getAnnotationsByType 메서드는 이 둘을 구분하지 않는다
(반복 가능 애너테이션, 그 컨테이너 애너테이션)** 따라서, 모두 가져온다

**(2) isAnnotationPresent 메서드는 둘을 명확히 구분한다.**

**(3) 따라서 반복 가능 애너테이션을 여러번 단 다음 isAnnotationPresent로 반복 가능 애너테이션이 달렸는지 검사한다면 “그렇지 않다”라고 알려준다
→ 컨테이너가 달렸기 때문에**

그 결과 애너테이션을 여러 번 단 메서드들을 모두 무시하고 지나친다.

같은 이유로 **isAnnotationPresent**로 컨테이너 애너테이션이 달렸는지 검사한다면 반복 가능 애너테이션을 한 번만 단 메서드를 무시하고 지나친다 → 컨테이너 애너테이션이 없기 때문에

> 반복 가능 애너테이션(`ExceptionTest`)을 여러 번 단다음 반복 가능 애너테이션이 있는지를 `isAnnotationPresent()`로 검사한다면 `false`를 반환한다.
>

> 이 경우 `Contatiner` 애너테이션(`ExceptionTestContainer`)을 인식하기 때문이다.
>

**그래서 달려있는 수와 상관없이 모두 검사하려면 둘을 따로따로 확인해야 한다.**

```java
**if (m.isAnnotationPresent(ExceptionTest.class)
|| m.isAnnotationPresent(ExceptionTestContainer.class)) {** **// 수정된 부분**
	tests++;
	try {
		m.invoke(null);
		System.out.printf("테스트 %s 실패: 예외를 던지지 않음%n", m);
	} catch (InvocationTargetException wrappedEx) {
			Throwable exc = wrappedEx.getCause();
			int oldPassed = passed;
			**ExceptionTest[] excType =**
				**m.getAnnotationsByType(ExceptionTest.class);  
				// 수정전 :** m.getAnnotation(ExceptionTest.class).value();
			**for (ExceptionTest excType : excTypes) {        // 수정된 부분**
				if (**excType.value()**.isInstance(exc)) {
					passed++;
					break;
				}
			}
			if (passed == OldPassed) 
				System.out.printf("테스트용S실패:85%n",m, exc);
		}
}
```

```java
public class Repeatable {

@ExceptionTest(IllegalArgumentException.class)
@ExceptionTest(NullPointerException.class)
publicvoidrepeatAnnotationMethod() {

    }

@ExceptionTest(IllegalArgumentException.class)
publicvoidnotRepeatAnnotationMethod() {

    }
}
```

> 반복 가능 애너테이션을 사용해 하나의 프로그램 요소에 같은 애너테이션을 여러 번 달 때의 코드 가독성을 높여보였다.
하지만 애너테이션을 선언하고 이를 처리하는 부분에서는 코드 양이 늘어나며, 특히 처리 코드가 복잡해져 오류가 날 가능성이 커질 수 있다.
>

다른 프로그래머가 소스코드에 추가 정보를 제공할 수 있는 도구를 만드는 일을 한다면 적당한 애너테이션 타입도 함께 정의해 제공하자.

## 애너테이션으로 할 수 있는 일을 명명 패턴으로 처리할 이유는 없다.

자바 프로그래머라면 예외 없이 자바가 제공하는 애너테이션 타입들은 사용해야 한다.
